tfhe/examples/sha256_bool/boolean_ops.rs

```diff
@@ -0,0 +1,394 @@
+        }).collect();
+
+        for (index, new_p, new_g) in updates {
+            propagate[index] = new_p;
+            generate[index] = new_g;
+        }
+
+        if d == 4 {
+            let mut cells = 0;
+            for d_2 in 0..4 { // last 4 stages
+                let stride = 1 << (4 - d_2 - 1);
+                cells += 1 << d_2;
+
+                let indices: Vec<(usize, usize)> = (0..cells).map(|cell| {
+                    (cell, stride + 2*stride*cell)
+                }).collect();
+
+                let updates: Vec<(usize, Ciphertext)> = indices.into_par_iter().map(|(_, index)| {
+                    let new_g = sk.or(&generate[index], &sk.and(&generate[index+stride], &propagate[index]));
+
+                    (index, new_g)
+                }).collect();
+
+                for (index, new_g) in updates {
+                    generate[index] = new_g;
+                }
+            }
+        }
+    }
+
+    let mut carry = trivial_bools(&[false; 32], sk);
+    for bit in 0..31 {
+        carry[bit] = generate[bit + 1].clone();
+    }
+
+    carry
+}
+
+// Implementation of the Ladner Fischer parallel prefix algorithm
+// This function may perform better than the previous one when many threads are available as it has less stages
+fn ladner_fischer(propagate: &[Ciphertext; 32], generate: &[Ciphertext; 32], sk: &ServerKey) -> [Ciphertext; 32] {
+    let mut propagate = propagate.clone();
+    let mut generate = generate.clone();
+
+    for d in 0..5 {
+        let stride = 1 << d;
+
+        let indices: Vec<(usize, usize)> = (0..32 - stride)
+            .rev()
+            .step_by(2 * stride)
+            .flat_map(|i| (0..stride).map(move |count| (i, count)))
+            .collect();
+
+        let updates: Vec<(usize, Ciphertext, Ciphertext)> = indices
+            .into_par_iter()
+            .map(|(i, count)| {
+                let index = i - count; // current column
+
+                let p = propagate[i + 1].clone(); // propagate from a previous column
+                let g = generate[i + 1].clone(); // generate from a previous column
+                let new_p;
+                let new_g;
+
+                if index < 32 - (2 * stride) { // black cell
+                    new_p = sk.and(&propagate[index], &p);
+                    new_g = sk.or(&generate[index], &sk.and(&g, &propagate[index]));
+
+                } else { // grey cell
+                    new_p = propagate[index].clone();
+                    new_g = sk.or(&generate[index], &sk.and(&g, &propagate[index]));
+                }
+                (index, new_p, new_g)
+            })
+            .collect();
+
+        for (index, new_p, new_g) in updates {
+            propagate[index] = new_p;
+            generate[index] = new_g;
+        }
+    }
+
+    let mut carry = trivial_bools(&[false; 32], sk);
+    for bit in 0..31 {
+        carry[bit] = generate[bit + 1].clone();
+    }
+
+    carry
+}
+
+// 2 (homomorphic) bitwise ops
```

Comment:
what does this mean ?

Reply:
I quantify the cost of each function using homomorphic bitwise operations as unit (kinda)